ibtk/include/ibtk/SnapshotCache.h

```diff
@@ -0,0 +1,226 @@
+
+#ifndef included_IBTK_SnapshotCache
+#define included_IBTK_SnapshotCache
+
+/////////////////////////////// INCLUDES /////////////////////////////////////
+
+#include <ibtk/config.h>
+
+#include <ibtk/ibtk_utilities.h>
+
+#include <tbox/Database.h>
+#include <tbox/Pointer.h>
+#include <tbox/Serializable.h>
+
+#include <CellVariable.h>
+#include <EdgeVariable.h>
+#include <FaceVariable.h>
+#include <HierarchyDataOpsReal.h>
+#include <IntVector.h>
+#include <NodeVariable.h>
+#include <PatchHierarchy.h>
+#include <SideVariable.h>
+
+#include <map>
+#include <set>
+#include <string>
+#include <vector>
+
+/////////////////////////////// CLASS DEFINITION /////////////////////////////
+
+namespace IBTK
+{
+/*!
+ * \brief Class SnapshotCache provides a method of storing snapshots of patch data on a snapshot of the patch hierarchy.
+ *
+ * This class can be used to store one type of variable, which is specified in the constructor.
+ *
+ * When retrieving or storing data with this class, the same patch data layout must be used, e.g. you must consistently
+ * store/retrieve CellData<NDIM, double> with depth of 2.
+ */
+class SnapshotCache : SAMRAI::tbox::Serializable
+{
+public:
+    /*!
+     * The constructor for class SnapshotCache sets some default values, reads in configuration information from input
+     * databases, and registers a variable/context pair with the VariableDatabase. Can optionally set up this class for
+     * restarts.
+     *
+     * The optinal input database is searched for the following keys
+     * - 'refine_type' : String that is the default refine operation for transfering data between hierarchies.
+     * - 'gcw' : Integer array that is the ghost cell width with which the internal variable is set up.
+     *
+     * \note If this object is being created from a restart, the grid geometry used the patch hierarchy must be
+     * supplied. An error will be thrown if the grid geometry is an invalid pointer.
+     */
+    SnapshotCache(std::string object_name,
+                  SAMRAI::tbox::Pointer<SAMRAI::hier::Variable<NDIM> > var,
+                  SAMRAI::tbox::Pointer<SAMRAI::tbox::Database> input_db = nullptr,
+                  bool register_for_restart = false,
+                  SAMRAI::tbox::Pointer<SAMRAI::hier::GridGeometry<NDIM> > grid_geom = nullptr);
+
+    /*!
+     * \brief The destructor for class SnapshotCache deallocates patch data as needed.
+     */
+    ~SnapshotCache();
+
+    /*!
+     * Clear the snapshots stored in this object. This deallocates all patch indices stored by this object.
+     */
+    void clearSnapshots();
+
+    /*!
+     * Create a snapshot. This creates a snapshot of the current hierarchy, and copies that data from this patch index
+     * into a patch index maintained by this object. If you want to overwrite a snapshot, you should call
+     * updateSnapshot() instead. Data is not allocated in this class until this function is called.
```

Comment:
If you call "set" twice (instead of "set" and then "update"), is there an error?

Reply:
Related questions:

1. If you want to capture several fields at the same time and on the same patch hierarchy, do you call "set" once and then "update" for the rest? What happens if you call "set" twice?
2. What happens if you want to capture fields at the same time but on different patch hierarchies? Does the implementation assume a single patch hierarchy configuration is used for each time point?

Reply:
This is a little tricky, because we are storing times as doubles, there's always a question of machine precision. I'll outline the possibilities:

1. If the user calls `setSnapshot` and the time point is **exactly** equal to a previous time point, then there should be no error, and the snapshot should be updated to whatever argument was provided by the function.
2. If the user calls `setSnapshot` and the time point is **different** (including machine precision), then an additional snapshot is taken and stored at that new time point.
3. If the user calls `updateSnapshot` and the time point is **within the specified tolerance**, then the snapshot at the found time point is updated.
4. If the user calls `updateSnapshot` and the time point is **not within the specified tolerance**, an error will occur.

Reply:
This class is only designed to store one snapshot of one variable at various points in time. It should be relatively straightforward to store one snapshot of multiple variables at various points in time. All that's required is wrapping `vector`'s around the objects and adding some logic to determine which variable to update.

Reply:
> This is a little tricky, because we are storing times as doubles, there's always a question of machine precision. I'll outline the possibilities:
> 
> 1. If the user calls `setSnapshot` and the time point is **exactly** equal to a previous time point, then there should be no error, and the snapshot should be updated to whatever argument was provided by the function.
> 2. If the user calls `setSnapshot` and the time point is **different** (including machine precision), then an additional snapshot is taken and stored at that new time point.
> 3. If the user calls `updateSnapshot` and the time point is **within the specified tolerance**, then the snapshot at the found time point is updated.
> 4. If the user calls `updateSnapshot` and the time point is **not within the specified tolerance**, an error will occur.

I think it makes sense to have both `setSnapshot` and `updateSnapshot` take an optional tolerance with a reasonable default (sqrt of machine precision?).

Reply:
> This class is only designed to store one snapshot of one variable at various points in time. It should be relatively straightforward to store one snapshot of multiple variables at various points in time. All that's required is wrapping `vector`'s around the objects and adding some logic to determine which variable to update.

I think it is fine to restrict to a single field.

One other thing --- does the interface allow us to look up the patch hierarchy that the snapshot was stored on?

Reply:
> I think it makes sense to have both setSnapshot and updateSnapshot take an optional tolerance with a reasonable default (sqrt of machine precision?).

Assuming the `PatchHierarchy` automatically deletes the data allocated when it's destroyed (which should happen), there will be functionally no difference between these calls. We can just have a single function `storeSnapshot` which has the behavior

1. If the user calls `storeSnapshot` and a stored time point is within the specified tolerance, then the snapshot at the found time point is updated.
2. If the user calls `storeSnapshot` and a stored time point is not within the specified tolerance, then an additional snapshot is taken and stored at that new time point.

> One other thing --- does the interface allow us to look up the patch hierarchy that the snapshot was stored on?

Not currently, but that's an easy function to add.